In [138]:
import psycopg2
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [139]:
connection = psycopg2.connect(
   dbname='project_sql',
   user='skillfactory',
   host='84.201.134.129',
   password='cCkxxLVrDE8EbvjueeMedPKt',
   port=5432
)

Задание 3.1 Сколько вакансий есть в базе?

In [140]:
query31 = f'''select 
         count(id)
         from public.VACANCIES
           
        '''
df = pd.read_sql_query(query31,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\211822708.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query31,connection)


count
0  49197

Задание 3.2 Сколько работодателей?

In [141]:
query32 = f'''select 
         count(id)
         from public.EMPLOYERS
           
        '''
df = pd.read_sql_query(query32,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\511464399.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query32,connection)


count
0  23501

Задание 3.3 Сколько регионов?

In [142]:
query33 = f'''select distinct 
         count(*)
         from public.AREAS
           
        '''
df = pd.read_sql_query(query33,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\3972623408.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query33,connection)


count
0   1362

Задание 3.4 Сколько сфер деятельности?

In [143]:
query34 = f'''select
         count(id)
         from public.INDUSTRIES
           
        '''
df = pd.read_sql_query(query34,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\466402999.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query34,connection)


count
0    294

Выводы по предварительному анализу данных:
в среднем(минимум один) работодатель предлагает 2 вакансий (49197 / 23501)
в среднем(минимум в одном) в регионе предлагается 36 вакансий (49197 / 1362)
в среднем(минимум от одной) от каждой индустрии(сферы деятельности) предлагается 167 вакансий (49197 / 294)
в среднем(минимум в одном) регионе находится 17 работодателей (23501 / 1362)
в среднем(минимум одна) каждая индустрия представлена 79 компанией (работодателем) (23501 / 294)

Задание 4.1 Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Выберите пятёрку лидеров по количеству вакансий:

In [144]:
query41 = f'''SELECT a.name area_name, 
    Count(*) AS vacancies_count
    FROM areas a
    JOIN  vacancies v ON a.id = v.area_id
    GROUP BY area_name
    ORDER BY vacancies_count DESC
    LIMIT 5
           
        '''
df = pd.read_sql_query(query41,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\2388069226.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query41,connection)


area_name  vacancies_count
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892

Задание 4.2 Посмотрим на зарплаты. У какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой?

In [145]:
query42 = f'''SELECT
    count(v.name) cnt
FROM public.VACANCIES v
WHERE salary_from IS NOT NULL 
or salary_to IS NOT NULL 
           
        '''
df = pd.read_sql_query(query42,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\2258574034.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query42,connection)


cnt
0  24073

Задание 4.3 Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого числа.

In [146]:
query43 = f'''SELECT
    AVG(salary_from) bottom_line,
    AVG(salary_to) upper_bound
FROM public.VACANCIES v 
           
        '''
df = pd.read_sql_query(query43,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\2037127397.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query43,connection)


bottom_line    upper_bound
0  71064.657901  110536.741923

Задание 4.4 Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Какая пара находится на втором месте по популярности?

In [147]:
query44 = f'''SELECT
    v.schedule,
    v.employment,
    count(*) cnt
FROM public.VACANCIES v
GROUP BY v.schedule, v.employment
order by cnt desc
limit 2
        '''
df = pd.read_sql_query(query44,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\1949663202.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query44,connection)


schedule        employment    cnt
0       Полный день  Полная занятость  35367
1  Удаленная работа  Полная занятость   7802

Задание 4.5 Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [148]:
query45 = f'''SELECT
    v.experience,
    count(v.name) cnt
FROM public.VACANCIES v
GROUP BY v.experience
order by cnt
        '''
df = pd.read_sql_query(query45,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\4155052520.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query45,connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

Выводы по детальному анализу вакансий

Чем крупнее населенный пункт (площадь, население), тем больше вакансий предлагается в нем.

Средне-максимальная ЗП выше средне-минимальной в полтора раза (110537.0 / 71065.0).

По распорядку и типу занятости явный лидер: "Полный день/Полная занятость", хотя на 'пятки уже наступает' "Удаленная работа" (пока в ~4.5 раза меньше).

Задание 5.1 Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [149]:
query51 = f'''SELECT
    e.name,
    count(v.name) cnt
FROM public.VACANCIES v
JOIN public.EMPLOYERS e ON v.employer_id = e.id
GROUP BY e.name
order by cnt desc
limit 5
        '''
df = pd.read_sql_query(query51,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\2924797891.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query51,connection)


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

Задание 5.2 Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей. Впишите его название в поле ниже в том виде, который вернул запрос.

In [150]:
query52 = f'''SELECT
    a.name, 
    Count(e.id) AS employers_cnt, 
    Count(v.id) AS vacancies_count
    FROM areas a 
    JOIN employers e ON a.id = e.area
    LEFT JOIN vacancies v ON a.id = v.area_id
    WHERE v.id IS NULL
    GROUP BY 1
    ORDER BY 3, 2 DESC
    LIMIT 10
        '''
df = pd.read_sql_query(query52,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\944504758.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query52,connection)


name  employers_cnt  vacancies_count
0                 Россия            410                0
1              Казахстан            207                0
2     Московская область             75                0
3     Краснодарский край             19                0
4     Ростовская область             18                0
5               Беларусь             18                0
6            Азербайджан             17                0
7  Нижегородская область             16                0
8   Республика Татарстан             16                0
9             Узбекистан             15                0

Задание 5.3 Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии.
Выберите максимальное значение из получившегося списка.

In [151]:
query53 = f'''select 
e.name, 
count(distinct v.area_id) 
from public.employers e 
left join public.vacancies v on e.id=v.employer_id 
group by e.id 
order by 2 desc
limit 3
        '''
df = pd.read_sql_query(query53,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\3025821198.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query53,connection)


name  count
0      Яндекс    181
1  Ростелеком    152
2  Спецремонт    116

Задание 5.4 Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.
Введите количество, которое вернул запрос, в поле ниже.

In [152]:
query54 = f'''select 
    Count(*) AS employers_without_industries_count
    FROM employers e
    LEFT JOIN employers_industries ei ON e.id = ei.employer_id
    WHERE ei.employer_id IS NULL
        '''
df = pd.read_sql_query(query54,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\2290062181.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query54,connection)


employers_without_industries_count
0                                8419

Задание 5.5 Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.
Введите в поле ниже название этой компании так же, как оно указано в результате запроса.

In [153]:
query55 = f'''SELECT e.name
    FROM employers e
    JOIN employers_industries ei ON e.id = ei.employer_id
    GROUP BY e.name
    HAVING Count(distinct ei.industry_id) = 4
    ORDER BY e.name
    LIMIT 3
        '''
df = pd.read_sql_query(query55,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\547106197.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query55,connection)


name
0  101 Интернет
1      21vek.by
2          2ГИС

Задание 5.6 С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».

In [154]:
query56 = f'''SELECT 
    Count(DISTINCT ei.employer_id) AS emp_dev_po
    FROM employers_industries ei
    JOIN industries i ON ei.industry_id = i.id
    WHERE i.name = 'Разработка программного обеспечения'
        '''
df = pd.read_sql_query(query56,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\1984040128.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query56,connection)


emp_dev_po
0        3553

Задание 5.7* Для компании «Яндекс» выведите список регионов-миллионников , в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании, собранных в этой таблице. Должна получиться выборка такого вида (приведён пример результата для компании SberTech):

Сколько строк получилось в вашей выборке?

In [155]:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
text = requests.get(url).text

table_with_cities = BeautifulSoup(text, 'html.parser').find('span', id='Список_городов-миллионеров')
table_with_cities = table_with_cities.parent.next_sibling

while table_with_cities.name != 'table':
    table_with_cities = table_with_cities.next_sibling

table_with_cities_trs = table_with_cities.find('tbody').find_all('tr')

million_cities = []
for row in table_with_cities_trs:
    tds = row.find_all('td')
    if len(tds) > 1:
        million_cities.append(tds[1].text.strip().replace('[a]', ''))

print('Найдены следующие города-миллионнеры (всего-{}):'.format(len(million_cities)))
print(*million_cities, sep='\n')

Найдены следующие города-миллионнеры (всего-16):
Москва
Санкт-Петербург
Новосибирск
Екатеринбург
Казань
Нижний Новгород
Красноярск
Челябинск
Самара
Уфа
Ростов-на-Дону
Омск
Краснодар
Воронеж
Пермь
Волгоград


Какой результат получился в строке Total?

In [156]:
where_million_cities = "'" + "','".join(million_cities) + "'"
where_company_name = "'Яндекс'"

query57 = (f'''
(SELECT a.name AS name, Count(*) AS cnt
FROM areas a
JOIN vacancies v ON a.id = v.area_id
JOIN employers e ON e.id = v.employer_id
WHERE a.name IN ({where_million_cities}) AND e.name = {where_company_name}
GROUP BY a.name
)
UNION
(
SELECT 'Total' AS name, Count(*) cnt
FROM areas a
JOIN vacancies v ON a.id = v.area_id
JOIN employers e ON e.id = v.employer_id
WHERE a.name IN ({where_million_cities}) AND e.name = {where_company_name}
)
ORDER BY cnt, name
''')
df = pd.read_sql_query(query57,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\742053927.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query57,connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4            Казань   25
5             Пермь   25
6    Ростов-на-Дону   25
7            Самара   26
8               Уфа   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

Выводы по анализу работодателей

Больше всех вакансии (почти в 4 раза от "второго" места) предлагает компания Яндекс, далее в ТОП-25 (с количеством предлагаемых вакансий от 100), входят довольно крупные компании (ИТ, Банки, Связь, Нефть), которым постоянно требуется новые кадры: развитие, возможная "текучка кадров", расширение бизнеса и т.д.

Наибольшее количество регионов у "троицы": Яндекс, Ростелеком, Спецремонт, это может быть связано с широкой сетью офисов(предоставления услуг) по регионам, что точно характерно для Ростелекома. Также возможно идет "сбор специалистов по регионом" с возможной удаленной работой или предложением постоянной релокации в ближайший офис компании

Компания "Яндекс" представлена(предлагает вакансии) в каждом из регионов-миллионников, практически пропорционально (кол-во вакансий) населению этого региона.

Задание 6.1 Сколько вакансий имеет отношение к данным

In [157]:
query61 = f'''SELECT 
    Count(*) AS ds_count
    FROM vacancies
    WHERE Lower(name) LIKE '%data%' OR Lower(name) LIKE '%данн%'
        '''
df = pd.read_sql_query(query61,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\1050663447.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query61,connection)


ds_count
0      1771

Задание 6.2 Сколько есть подходящих вакансий для начинающего дата-сайентиста?

Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:

‘data scientist’;
‘data science’;
‘исследователь данных’;
‘ML’ (здесь не нужно брать вакансии по HTML);
‘machine learning’;
‘машинн%обучен%’.
Считаем вакансиями для специалистов уровня Junior следующие:

в названии есть слово “junior” или
требуемый опыт — «Нет опыта» или
тип трудоустройства — «Стажировка».
В качестве ответа запишите количество вакансий, которое вернул запрос.

In [158]:
data_sience_where = '''(
        Lower(name) LIKE '%data scientist%'
        OR Lower(name) LIKE '%data science%'
        OR Lower(name) LIKE '%исследователь данных%'
        OR Lower(name) LIKE '%machine learning%'
        OR Lower(name) LIKE '%машинн%обучен%%'
        OR (
            name LIKE '%ML%'
            AND name NOT LIKE '%HTML%'
        )
)'''

query62 = f'''SELECT 
    Count(*) AS junior_cnt
    FROM vacancies
    WHERE {data_sience_where}
        AND (
            Lower(name) LIKE '%junior%'
            OR experience = 'Нет опыта'
            OR employment = 'Стажировка'
        )
        '''
df = pd.read_sql_query(query62,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\2606212339.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query62,connection)


junior_cnt
0          51

Задание 6.3 Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

В качестве ответа запишите количество, которое вернул запрос.

In [159]:
query63 = f'''SELECT Count(*) AS ds_and_db_cnt
    FROM vacancies
    WHERE {data_sience_where}
        AND (
            Upper(key_skills) LIKE '%SQL%'
            OR Lower(key_skills) LIKE '%postgres%'
        )
        '''
df = pd.read_sql_query(query63,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\1474287407.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query63,connection)


ds_and_db_cnt
0            201

Задание 6.4 С помощью запроса, аналогичного предыдущему, проверьте, насколько популярен Python в требованиях работодателей к DS. Вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

В качестве ответа запишите количество, которое вернул запрос.

In [160]:
query64 = f'''SELECT 
    count(*) ds_and_python_cnt
    FROM vacancies
    WHERE {data_sience_where}
    AND  Lower(key_skills) LIKE '%python%'
        '''
df = pd.read_sql_query(query64,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\3412962477.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query64,connection)


ds_and_python_cnt
0                351

Задание 6.5 Сколько ключевых навыков в среднем указывают в вакансиях для DS?

Ответ округлите до двух знаков после точки-разделителя.

In [161]:
query65 = f'''SELECT 
        Round(
        Avg(1 + length(key_skills) - length(replace(key_skills, chr(9), ''))), 2) AS ds_key_skills_cnt
    FROM vacancies
    WHERE {data_sience_where}
    AND  key_skills IS NOT NULL
        '''
df = pd.read_sql_query(query65,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\3381444444.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query65,connection)


ds_key_skills_cnt
0               6.41

Задание 6.6 Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля experience).
При решении задачи примите во внимание следующее:

Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
Если заполнены оба поля с зарплатой, считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, его и считаем зарплатой по вакансии.
Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null).
Чтобы избежать этой ситуации, мы воспользуемся функцией coalesce , которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос select 1 + coalesce(null, 0).
Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [162]:
query66 = f'''SELECT 
        experience,
        Round(Avg((Coalesce(salary_from, salary_to) + Coalesce(salary_to, salary_from)) / 2)) AS ds_avg_salary
    FROM vacancies
    WHERE  (
            salary_to IS NOT NULL
            OR salary_from IS NOT NULL
           ) 
        AND {data_sience_where}  
    GROUP BY experience
        '''
df = pd.read_sql_query(query66,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\4194555082.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query66,connection)


experience  ds_avg_salary
0           Нет опыта        74643.0
1  От 1 года до 3 лет       139675.0
2       От 3 до 6 лет       243115.0

Выводы по предметному анализу

Вакансии конкретно для DataSience занимают примерно 3.6% (1771 / 49197 * 100) от общего числа вакансий, что в общем-то не мало.
Среди представленных вакансий найдется место и юниорам 2.9% (51 / 1771 * 100)

Каких вакансий много?

In [163]:
query7 = f'''SELECT 
    name AS vacancy_name, 
    Count(*) AS vacancy_count
    FROM vacancies
    GROUP BY 1
    ORDER BY 2 DESC
    LIMIT 50
        '''
df = pd.read_sql_query(query7,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\3634142481.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query7,connection)


vacancy_name  vacancy_count
0                             Системный администратор           1761
1                                      Программист 1С           1153
2                                     Бизнес-аналитик            488
3                    Специалист технической поддержки            457
4                                  Системный аналитик            432
5                                            Аналитик            404
6                                 Инженер-программист            389
7                                      Программист 1C            360
8                                Frontend-разработчик            242
9                       Инженер технической поддержки            230
10                                        Программист            224
11                               Руководитель проекта            191
12            Младший специалист ручного тестирования            169
13        Специалист службы поддержки в Яндекс Бизнес            168
14                                  Менеджер проектов            153
15                                   Java-разработчик            150
16          Специалист по информационной безопасности            148
17  Специалист службы поддержки Рекламной сети Янд...            139
18                                    PHP-разработчик            135
19          Специалист службы поддержки пользователей            126
20                                Компьютерный мастер            123
21                                   Менеджер проекта            120
22                               Технический писатель            111
23                                Маркетолог-аналитик            108
24                              Руководитель проектов            106
25                             Ведущий программист 1C            106
26                             Технический специалист            104
27                 Сотрудник поддержки поиска Яндекса             99
28         Технический писатель (дистанционный автор)             99
29                    Ведущий системный администратор             99
30                Специалист поддержки в Yandex Cloud             97
31                             Ведущий программист 1С             94
32                               Технический директор             94
33                        Ведущий инженер-программист             92
34                                        QA Engineer             91
35                                    Web-разработчик             88
36                                    IOS разработчик             86
37                 Помощник системного администратора             86
38                            Программист-разработчик             85
39                                        Аналитик 1С             85
40  Специалист службы поддержки с техническими зна...             84
41                                     Разработчик 1С             83
42                                Компьютерный Мастер             83
43                                     Консультант 1С             82
44                                    Project Manager             81
45                                    Web-программист             81
46                       Ученик компьютерного мастера             79
47                                    Аналитик данных             78
48                                            Инженер             78
49                                    Сетевой инженер             78

Как показывает запрос выше в ТОП-50 входят вакансии, связанные с ИТ, среди них пара лидеров: "Системный администратор" и "Программист 1С"

Что можно сказать о рынке труда в различных городах?

In [164]:
query8 = f'''SELECT a.name AS area_name,
            Round(Avg((Coalesce(salary_from, salary_to) + Coalesce(salary_to, salary_from)) / 2)) AS ds_avg_salary, 
            Count(v.id) AS vacancies_count
    FROM areas a
    JOIN vacancies v ON v.area_id = a.id
    WHERE  (
            v.salary_to   IS NOT NULL OR
            v.salary_from IS NOT NULL
        )
    GROUP BY area_name
    HAVING Count(v.id) > 10
    ORDER BY ds_avg_salary DESC
    LIMIT 20
        '''
df = pd.read_sql_query(query8,connection)
df

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\1297704090.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query8,connection)


area_name  ds_avg_salary  vacancies_count
0            Турция       235435.0               19
1              Кипр       223736.0               39
2           Армения       219570.0              106
3            Сербия       208423.0               12
4           Тбилиси       207156.0               80
5            Батуми       199834.0               18
6         Иннополис       175882.0               17
7              Баку       155747.0               51
8            Москва       134140.0             1592
9             Лобня       121136.0               11
10          Магадан       117033.0               19
11       Зеленоград       116946.0               76
12      Новосибирск       107777.0              929
13  Санкт-Петербург       104410.0             1203
14             Сочи        99437.0              155
15            Томск        99319.0              350
16            Химки        98641.0               51
17  Нижний Новгород        96059.0              680
18            Пермь        95985.0              352
19           Мытищи        94567.0               60

Более высокие (средние) ЗП в города за пределами России.

Какие ещё интересные исследования можно провести?

ЗП дата-сайентистов, неплохая медиана и очень хорошая ЗП у специалистов.

In [172]:
query9 = f'''SELECT  (Coalesce(salary_from, salary_to) + Coalesce(salary_to, salary_from)) / 2 AS salary
    FROM vacancies
    WHERE  (
            salary_to   IS NOT NULL OR
            salary_from IS NOT NULL
        ) AND {data_sience_where} 
'''
df = pd.read_sql_query(query9,connection)
df.median()

C:\Users\Acer\AppData\Local\Temp\ipykernel_11296\3904633403.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query9,connection)


salary    162500.0
dtype: float64